plugins/out_es/es_conf.c

```diff
@@ -194,88 +196,190 @@ struct flb_elasticsearch *flb_es_conf_create(struct flb_output_instance *ins,
+  if (ret == -1) {
+      if (ec) {
+          flb_es_conf_destroy(ec);
+      }
+      return -1;
+  }
+
+    return 0;
 }
 
-int flb_es_conf_destroy(struct flb_elasticsearch *ctx)
+/* Configure in HA mode */
+int es_config_ha(const char *upstream_file,
+                 struct flb_elasticsearch *ctx,
+                 struct flb_config *config)
 {
-    if (!ctx) {
-        return 0;
+    ssize_t ret = 0;
+    const char *tmp;
+    const char *path;
+    struct mk_list *head;
+    struct flb_uri *uri = ctx->ins->host.uri;
+    struct flb_uri_field *f_index = NULL;
+    struct flb_uri_field *f_type = NULL;
+    struct flb_upstream_node *node;
+    struct flb_elasticsearch_config *ec = NULL;
+
+    ctx->ha_mode = FLB_TRUE;
+    ctx->ha = flb_upstream_ha_from_file(upstream_file, config);
+    if (!ctx->ha) {
+        flb_plg_error(ctx->ins, "cannot load Upstream file");
+        return -1;
+    }
+
+    if (uri) {
+        if (uri->count >= 2) {
+            f_index = flb_uri_get(uri, 0);
+            f_type  = flb_uri_get(uri, 1);
+        }
+    }
+
+    /* Iterate nodes and create a forward_config context */
+    mk_list_foreach(head, &ctx->ha->nodes) {
+        node = mk_list_entry(head, struct flb_upstream_node, _head);
+
+        /* Allocate context */
+        ec = flb_calloc(1, sizeof(struct flb_elasticsearch_config));
+        if (!ec) {
+            flb_errno();
+            flb_plg_error(ctx->ins, "failed config allocation");
+            continue;
+        }
+
+        /* Set manual Index and Type */
+        if (f_index) {
+            ec->index = flb_strdup(f_index->value); /* FIXME */
+        }
+
+        if (f_type) {
+            ec->type = flb_strdup(f_type->value); /* FIXME */
+        }
+
+        /* Set default values */
+        ret = flb_output_config_map_set(ctx->ins, ec);
+        if (ret == -1) {
+            flb_free(ec);
+            return -1;
+        }
+
+        /* Elasticsearch: Path */
+        path = flb_upstream_node_get_property("path", node);
+        if (!path) {
+            path = "";
+        }
+
+        /* Elasticsearch: Pipeline */
+        tmp = flb_upstream_node_get_property("pipeline", node);
+        if (tmp) {
+            snprintf(ec->uri, sizeof(ec->uri) - 1, "%s/_bulk/?pipeline=%s", path, tmp);
+        }
+        else {
+            snprintf(ec->uri, sizeof(ec->uri) - 1, "%s/_bulk", path);
+        }
+
+        /* Initialize and validate es_config context */
+        ret = flb_es_conf_init(ec, ctx);
+        if (ret == -1) {
+            if (ec) {
+                flb_es_conf_destroy(ec);
+            }
+            return -1;
+        }
+
+        /* Set our elasticsearch_config context into the node */
+        flb_upstream_node_set_data(ec, node);
     }
 
-    if (ctx->u) {
-        flb_upstream_destroy(ctx->u);
+    return 0;
+}
+
+int flb_es_conf_init(struct flb_elasticsearch_config *ec,
+                     struct flb_elasticsearch *ctx)
+{
+    mk_list_add(&ec->_head, &ctx->configs);
```

Comment:
I have a strong feeling that something is missing here.

Especially the current patch ignores the custom configuration in `[NODE]`
e.g. `Buffer_Size` in the following example is ignored.

```ini
[UPSTREAM]
    name       forward-balancing

[NODE]
    Name       node-1
    Host       127.0.0.1
    Port       9200
    Buffer_Size 4096
```

I assume you have intended to implement that functionality in
this function but forgotten to push the actual chageset?